File: scripts/safety-net.lua
local forced_heal_percent = 30

local function active_attackers(player)
  local attackers = {}
  for _, npc in ipairs(gc.read("npcs", { within = 16, limit = 40 })) do
    if npc.interacting == player.name then attackers[#attackers + 1] = npc end
  end
  return attackers
end

local function emergency_hitpoints(player)
  return player.current_hitpoints > 0 and player.max_hitpoints > 0 and
    player.current_hitpoints * 100 < player.max_hitpoints * forced_heal_percent
end

local function has_action(item, expected)
  for _, action in ipairs(item.actions or {}) do
    if action == expected then return true end
  end
  return false
end

local function eat_available_food()
  for _, item in ipairs((gc.read("inventory") or {}).items or {}) do
    if has_action(item, "Eat") then
      local receipt = gc.await {
        action = { type = "item.interact", id = item.id, action = "Eat" },
        timeout = { game_ticks = 6 },
        policy = { breaks = false, cursor_release = "none", fidget = "none" },
      }
      if receipt.status == "dispatched" or receipt.status == "complete" then
        return receipt
      end
    end
  end
  return { status = "rejected", result = "no_edible_inventory_item" }
end

local function completed_escape(receipt)
  return receipt.result == "emergency_escape_complete" or
    receipt.result == "emergency_food_and_escape_complete"
end

return {
  run = function()
    gc.activity("manual")
    gc.state("attention_required")
    while true do
      local player = gc.read("player")
      local attackers = active_attackers(player)

      if #attackers > 0 and emergency_hitpoints(player) then
        gc.state("recovering_active_combat")
        gc.overlay {
          { label = "Safety Net", value = "Recovering from active combat" },
          { label = "Attacker", value = attackers[1].name or tostring(attackers[1].id) },
          { label = "HP", value = player.current_hitpoints .. "/" .. player.max_hitpoints },
        }

        local recovered = gc.await {
          action = { type = "safety.recover" },
          timeout = { game_ticks = 6 },
          policy = { breaks = false, cursor_release = "none", fidget = "none" },
        }
        if completed_escape(recovered) then
          return {
            status = "complete",
            result = "safety_net_active_combat_escape",
            recovery_attempted = true,
            recovery_completed = true,
            receipt = recovered,
          }
        end

        local controller_handled_food = recovered.result == "emergency_consumable_dispatched" or
          recovered.result == "safety_recovery_already_running"
        if not controller_handled_food then eat_available_food() end
      end

      gc.state("attention_required")
      gc.overlay {
        {
          label = "Safety Net",
          value = "Awaiting manual control after script failure",
        },
        {
          label = "HP",
          value = player.current_hitpoints .. "/" .. player.max_hitpoints,
        },
      }
      gc.await { event = "game.tick" }
    end
  end,
}

File: src/main/java/com/genericclient/scripts/recovery/SafetyNet.java
package com.genericclient.scripts.recovery;

import com.genericclient.script.Automation;
import com.genericclient.script.ScriptScope;
import com.genericclient.script.ScriptSettings;
import com.genericclient.script.SnapshotData;
import java.util.Map;
import org.dreambot.api.Client;
import org.dreambot.api.methods.container.impl.Inventory;
import org.dreambot.api.script.AbstractScript;
import org.dreambot.api.script.Category;
import org.dreambot.api.script.ScriptManifest;
import org.dreambot.api.wrappers.items.Item;

@ScriptManifest(name="Safety Net",author="GenericClient",category=Category.UTILITY,version=1,
	description="Wait for manual control after a failure and retain emergency recovery during active combat.")
@ScriptSettings(id="safety-net")
public final class SafetyNet extends AbstractScript
{
	@Override public int onLoop()
	{
		if (!Client.isLoggedIn()) return 600;
		Automation.activity("manual");
		String name = com.genericclient.scripts.shared.WorkflowScript.player().getName();
		boolean attacked = SnapshotData.rows("npcs",Map.of("within",16,"limit",40)).stream()
			.anyMatch(npc -> name.equals(npc.get("interacting")));
		if (attacked && com.genericclient.scripts.shared.WorkflowScript.player().getHealthPercent() < 30)
		{
			Map<String,Object> result = ScriptScope.current().execute("safety.recover",Map.of(),60_000);
			String outcome = String.valueOf(result.get("result"));
			if (outcome.equals("emergency_escape_complete") || outcome.equals("emergency_food_and_escape_complete"))
			{
				Automation.finish(Map.of("status","complete","result","safety_net_active_combat_escape"));
				return -1;
			}
			// A configured guard owns the meal through its approved consumables, so the safety net eats only when no guard exists.
			if (outcome.equals("safety_net_not_configured"))
			{
				Item food = Inventory.get(item -> item.hasAction("Eat"));
				if (food != null) food.interact("Eat");
			}
		}
		Automation.overlay(Map.of("Safety Net","Awaiting manual control","HP",com.genericclient.scripts.shared.WorkflowScript.player().getHealthPercent() + "%"));
		return 600;
	}
}
